### PR TITLE
play against computer

### LINKED ISSUE #12: Play against computer
Add a button which will allow user to play against computer. In this feature the user and the computer will take turns after each other.

### PR DESCRIPTION
**Pull request by MuhammadAwaisGondal**
I have added a toggle button to choose the game mode. User can either play with computer or with other player. When playing against the computer the user will take the 1st turn and the computer will take next turn. Computer's turn will be calculated by a function named findBestMove which check the available squares and randomly select one empty square.

### CODE REVIEW

**Comment by rea-fernandes** [comment]
Please add the description of changes in this pull request.

File: src/Game/Game.js
Comment (rea-fernandes) [inline src/Game/Game.js:1]: The computer version always fills the next free spot starting from the top. It is better to fill in randomly or best approach is to check if other player has 2 continuous blocks filled.

File: src/Game/Game.js [hunk 1]
```diff
@@ -83,12 +92,30 @@ export default function Game() {
...
+        </button>
+      </div>
       <div className="game-board">
         <Board xIsNext={xIsNext} squares={currentSquares} onPlay={handlePlay} />
       </div>
       <div className="game-info">
+        <div className="status">
+          {calculateWinner(currentSquares)
+            ? "Winner of this round is: " + calculateWinner(currentSquares)
+            : `Next player is: ${xIsNext ? "O" : "X"}`}
```
Comment (rea-fernandes) [inline src/Game/Game.js]: image

1. The announcement is duplicated.
2. Add a break between Toggle player button and announcement.

File: src/Game/Game.js [hunk 2]
```diff
@@ -56,13 +55,23 @@ function Board({ xIsNext, squares, onPlay }) {
 export default function Game() {
   const [history, setHistory] = useState([Array(9).fill(null)]);
   const [currentMove, setCurrentMove] = useState(0);
+  const [playAgainstComputer, setPlayAgainstComputer] = useState(false);
```
Comment (MuhammadAwaisGondal) [inline src/Game/Game.js:58]: the state variable to set the game against computer on other player

File: src/Game/Game.js [hunk 3]
```diff
@@ -56,13 +55,23 @@ function Board({ xIsNext, squares, onPlay }) {
...
     setCurrentMove(nextHistory.length - 1);
+
+    if (playAgainstComputer && !calculateWinner(nextSquares)) {
+      const computerMove = findBestMove(nextSquares);
+      const newSquares = nextSquares.slice();
+      newSquares[computerMove] = "X";
+
+      setHistory([...nextHistory, newSquares]);
+      setCurrentMove(nextHistory.length);
+    }
```
Comment (MuhammadAwaisGondal) [inline src/Game/Game.js:74]: This code is checking if the computer is playing and then taking the computer turn by calling findBestMove function

File: src/Game/Game.js [hunk 4]
```diff
@@ -83,12 +92,30 @@ export default function Game() {
     );
   });
 
+  const handleTogglePlayer = () => {
+    setPlayAgainstComputer(!playAgainstComputer);
+    setHistory([Array(9).fill(null)]);
+    setCurrentMove(0);
+  };
+
```
Comment (MuhammadAwaisGondal) [inline src/Game/Game.js:100]: Toggle function to change between computer and human game

File: src/Game/Game.js [hunk 5]
```diff
@@ -83,12 +92,30 @@ export default function Game() {
...
   return (
     <div className="game">
+      <div className="game-top">
+        <button onClick={handleTogglePlayer}>
+          {playAgainstComputer
+            ? "Play against other player"
+            : "Play against computer"}
+        </button>
+      </div>
```
Comment (MuhammadAwaisGondal) [inline src/Game/Game.js:109]: Toggle button

File: src/Game/Game.js [hunk 6]
```diff
@@ -114,3 +141,12 @@ function calculateWinner(squares) {
...
 }
+
+function findBestMove(squares) {
+  for (let i = 0; i < squares.length; i++) {
+    if (!squares[i]) {
+      return i;
+    }
+  }
+  return -1;
+}
```
Comment (MuhammadAwaisGondal) [inline src/Game/Game.js:152]: This function find the move for computer

File: src/Game/Game.js [hunk 7]
```diff
@@ -72,9 +81,9 @@ export default function Game() {
   const moves = history.map((squares, move) => {
     let description;
     if (move > 0) {
-      description = 'Go to move #' + move;
+      description = "Go to move #" + move;
```
Comment (rea-fernandes) [inline src/Game/Game.js:85]: There is a bug in play against computer mode when go to move is selected.

Comment (rea-fernandes) [reply src/Game/Game.js:85]: image

Human is O and Computer is X, but with Go to move button, Human can play computer's move.

**Review by yalihuss — APPROVED** [review]

**Review by rea-fernandes — APPROVED** [review]

**Pull request merged**